File: 01/js/01.js
#!/usr/bin/env node

const fs = require('fs');

fs.readFile('input', 'utf-8', (_, s) => {
    const input = s.split('\n').map(x => +x),
        f = (sum, remaining, idx) => remaining ?
            input.slice(idx).map((n, i) => n * f(sum+n, remaining-1, i+1)).find(x => x) :
            sum == 2020;
    // both parts
    [2,3].map(n => console.log(f(0, n)));
});

File: 01/hs/01.hs
#!/usr/bin/env runhaskell

import Data.List
import Data.Maybe

comb _ 0 = [[]]
comb [] _ = []
comb (x:xs) n = map (x:) (comb xs (n-1)) ++ comb xs n

go = ((product . fromJust . find ((==2020) . sum)) .) . comb

main = do
    input <- map read . lines <$> readFile "input"
    -- both parts
    mapM (putStrLn . show . go input) [2,3]
